### ISSUE TITLE
New Maintainer Wanted :-)

### ISSUE THREAD

**Issue by maelle** [ropensci/essurvey#57]
Or new maintainer team. :smile_cat: 

**Because of #56 a whole overhaul of the package is needed**

If you're interested, please comment in the issue.
For more info, see
- [our guidance on Changing package maintainers](https://devguide.ropensci.org/changing-maintainers.html);
- [our Package Curation Policy](https://devguide.ropensci.org/curationpolicy.html).

**Comment by maelle** [comment]
@ropensci/admin

**Comment by etiennebacher** [comment]
Hello, I can't say that I will have time or knowledge to be maintainer but I have used `essurvey` in the past so I'd like to contribute back to it.

However, before thinking whether I should be a maintainer, I'm not even sure it's technically possible to write a package that automatically downloads the ESS datasets anymore. First, there's no API. Second, from what I can see from the current code, the old way was "easy" in the sense that each dataset had its own fixed URL to download the data. However, it seems that it is no longer the case. Indeed, all "Download" buttons now lead to the same URL: https://ess-search.nsd.no/en/download

This means that it is not possible to distinguish one dataset from the other based on the URL. Maybe we could perform the POST request to trigger the download ourselves using `httr`/`httr2`, but it seems that the parameters passed in the POST request are random and change at each download:

![image](https://user-images.githubusercontent.com/52219252/196217971-fede8d59-6024-4c62-bf17-b31568ffc954.png)

There is also a `GET` request made but the parameters are also random. 

Therefore, it seems to me that the only way to bring `essurvey` back to life is to contact directly the organization that manages it to ask if they have some plans to make an API, or at least make it possible to programmatically download the datasets. But maybe I'm wrong and there's a way to do it. If someone finds a solution and doesn't have time to implement it, I could try to integrate it in the package.

**Comment by maelle** [comment]
Wow, thanks a lot @etiennebacher for the digging! So yes, I suppose a new maintainer would need to contact the data provider first.

**Comment by cimentadaj** [comment]
@djhurio can help as he has contact with the ESS organizing team. @etiennebacher, you're right as we're a bit lost on how the data is now being downloaded. I haven't looked into how it works now but you're on the right track.

**Comment by gorcha** [comment]
Hi all,

@etiennebacher it looks like programmatically accessing metadata and downloading data can be done on the new data portal using a GraphQL API, but data downloads need a few steps. There are API docs [here](https://docs.nsd.no/) but they're not particularly illuminating, I've just been following the request flow in dev tools to see what it does. It looks a bit fiddly but very doable.

It also needs an OAuth2 authentication flow, which is a bit of a pain. I can't see any option to register an OAuth2 client app through the ESS data portal so that would probably need help from the ESS organising team.

I'm a bit time-strapped at the moment but would be more than happy to help get something up and running!

**Comment by maelle** [comment]
Thanks @gorcha!! Should I go ahead and give you write access to this repository? (if you decide to become the maintainer, you'd get admin access)

**Comment by etiennebacher** [comment]
@gorcha I see the GraphQL queries, but I don't know how to reproduce them. Some elements look random, like the `datafileId`. As you said, the NSD API docs are not helpful. 

For reference, here are the steps I follow:
* go to https://ess-search.nsd.no/
* go to "ESS Round 10"
* click on the "Download" button
* choose any format, for example "SAV"
* start download

Here's the first GraphQL query for me:

![image](https://user-images.githubusercontent.com/52219252/197516084-fd37d720-96f9-4306-82c3-83af85e28973.png)

And the second:

![image](https://user-images.githubusercontent.com/52219252/197516164-6553fd67-c27a-4128-a448-056fe05f3099.png)

I have more or less the same requests when I use the "Data Wizard". The only difference is that there are more arguments because we specify which variables/years/countries we want. I don't know how to mimic this from R. If you have an example, I'd be curious to see how you do it.

**Comment by gorcha** [comment]
Hi @maelle - sure! I won't have a chance to look at it for a few weeks though.

Hi @etiennebacher, the datafile IDs are retrieved as part of some earlier GraphQL queries. In addition to watching the requests I've been prodding the JavaScript a bit to figure out some of the details.

Here's a dump of the process flow for both the Data Portal and Data Wizard from a bit of poking around, with the GraphQL operation name and arguments for each step. One note - `instance` and `agencyId` are used a lot, and from what I can see they are always `"PUBLISHED"` and `"INT_ESSERIC"` respectively.

I haven't had a go at replicating it in R yet, but the GraphQL calls are all just JSON so there'll be some parsing shenanigans and dealing with HTTP responses but the trickiest part will probably be OAuth.

## Data Portal

#### Study metadata

graphql: `query studyMetadata($id: ID!, $instance: Instance!, $agencyId: Agency!)`

Retrieve metadata for the study (e.g. ESS 2010), which is used to populate the page (the study ID is in the page URL). The set of datafiles for the study are part of the returned object, including the datafile ID, description etc.

#### Get download URL

graphql: `query download($datafileId: ID!, $version: Int!, $agencyId: String!, $format: DownloadFormat!, $instance: String!)`

Get the download URL. This uses the datafileID, version, agencyId and instance values returned from `studyMetadata`. format is "CSV", "SAV" etc.


#### Download

`https://stessdissprodwe.blob.core.windows.net/data/download/2/download_dataset_*`

Download the data set from the URL returned by `download`

#### Register the download

graphql: `mutation registerSingleDownload($datafileUrn: String!, $datafileTitle: String!, $studyUrn: String!, $format: DownloadFormat!, $orderedBy: ID!)`

This registers downloads so they show up in the "previous downloads" section (not sure if there is any other purpose for this). `datafileUrn` and `studyUrn` are created from metadata using the datafile/study ID, version and agency (agency is always `int.esseric`):
```js
'urn:ddi:int.esseric:' + t.id + ':' + t.version
```

`format` is CSV, SAV, etc., and `orderedBy` is the ID of the current user (which is retrieved via a separate couple of graphql queries).

## Data Wizard

#### Get variable metadata

graphql: `query topLevelConceptualVariablesGroups($id: ID!, $version: Int)`
graphql: `query variableMetadata($id: ID!, $version: Int, $instance: Instance!, $agencyId: Agency!)`
graphql: `query codeComparisonTable($input: CodeComparisonInput!)`

Variable metadata (high level conceptual groups, variable details) are retrieved for the entire ESS series. This works similarly to the study metadata/data files - the top level conceptual variable groups are accessed from the series metadata (using the series ID). The returned object contains variable details (ID, name, etc.), and further variable info is accessed through additional graphql queries.

#### Get country metadata

graphql: `query countryCoverageTable($id: ID!, $version: Int)`

Possible country combinations are returned as a "countryCoverageTable" object (based on the series ID) that contains datafile IDs, country IDs, and a flag telling us whether the country exists for each datafile.

#### Make the wizard generate the file

graphql: `query ($input: WizardDownloadInput!)`

This asks the wizard to create the file. The WizardDownloadInput tells it what datafiles, countries and variables to include and ties it to the current user with `orderedBy`. For e.g.
```json
{
  "variables": {
    "input": {
      "agencyId": "INT_ESSERIC",
      "datafiles": [
        {
          "countries": null,
          "id": "ffc43f48-e15a-4a1c-8813-47eda377c355",
          "version": 73
        },
        {
          "countries": [
            "AT"
          ],
          "id": "b2b0bf39-176b-4eca-8d26-3c05ea83d2cb",
          "version": 248
        }
      ],
      "format": "SAV",
      "instance": "PUBLISHED",
      "variables": [
        "netuse",
        "netusoft",
        "netustm",
        "nwspol",
        "nwsppol",
        "nwsptot",
        "pplfair",
        "pplhlp",
        "ppltrst",
        "rdpol",
        "rdtot",
        "tvpol",
        "tvtot"
      ]
    }
  }
}
```

The response gives the ID for the wizard download to be used in the following steps.

#### Register the download

graphql: `mutation registerWizardDownload($datafiles: [WizardDataFileDownloadInput!]!, $variables: [String!]!, $orderedBy: ID!, 
$format: DownloadFormat!)`

Similar to the Data Portal download registration, but I can't see this being used in the UI anywhere (there are no previous downloads displayed for the data wizard).

#### Poll the wizard download

While the file is being generated/prepared it is repeatedly polled to see if it's ready yet. This returns the an isFinished indicator and the URL for the download.

graphql: `query ($id: ID!) {
  ESS {
    pollWizardDownload(id: $id) {
      isFinished
      url
      __typename
    }
    __typename
  }
}
`

#### Download

`https://stessdissprodwe.blob.core.windows.net/data/download/2/start_generate_datafile_job_*`

Once the `pollWizardDownload` returns `isFinished: true`, data is downloaded from the URL.

**Comment by etiennebacher** [comment]
Thanks a lot for all these explanations! I don't have the skills to make this work, I never used GraphQL before so I've no idea where to start. I'll try to help in other ways

**Comment by maelle** [comment]
@gorcha I've now invited you to the ropensci organization and to a team with write access to this repository. Note that you'll need to enable 2FA for your GitHub account if that's not already the case, see https://docs.github.com/en/authentication/securing-your-account-with-two-factor-authentication-2fa/configuring-two-factor-authentication + https://ropensci.org/blog/2022/05/16/requiring-2fa-for-the-ropensci-github-organization/ for context

@etiennebacher if you start contributing (thanks already for the convo here!!) please ping me so I might grant you access too.

**Comment by gorcha** [comment]
Thanks @maelle!

No worries @etiennebacher, any help with testing and documentation would be super helpful 🙂
I'll ping you once I get started if there's anything specific. Thanks!

**Comment by LukasWallrich** [comment]
@gorcha Just to say that I'd be happy to help as well. I've loved using this package in both teaching and research and would be glad to see it brought back to life :)

**Comment by LukasWallrich** [comment]
Just to throw that out there: I now saw that the ESS data is CC BY-NC-SA 4.0 ... and not very large (10 waves of apparently only ~15 MBs). So an alternative to OAuth and API issues might just be to host the data in a separate GitHub repo, accessed by the package?

**Comment by gorcha** [comment]
Hey @LukasWallrich,

Great idea, I hadn't even thought of that! Will check it out :)

**Comment by djhurio** [comment]
Dear all, thank you for your involvement into this issue with the ESS data. Regarding the idea by @LukasWallrich, please note the ESS data for each round is released in several releases and versions (data editions are numbered according `release.version` scheme up to my knowledge). Usually for reach round there are two to three data releases, where each new data release contains data from more countries than previous release. Each release can have several versions because of corrections made to the data. For example we can see there were three data releases for the round 9 data and currently only the first release of the round 10 data has been made available (more releases are in pipeline).

```
   essround edition   proddate     N
1:      R07     2.2 2018-12-01 40185
2:      R08     2.2 2020-12-10 44387
3:      R09     3.1 2021-02-17 49519
4:      R10     1.2 2022-06-28 18060
```

If you host the data on a separate repo than there should be a process to monitor if new data edition (`release.version`) for each round has been published by the ESS.

**Comment by etiennebacher** [comment]
Hello, any news about this?

@gorcha did you have the occasion to try making the graphql requests from R? Or is the plan to use a separate repo to host the data and keep it updated with future releases?

---

FYI, I contacted the organization that manages the ESS data. They said that providing a clear API documentation is something they want to do but they didn't give me a timeline for this.

**Comment by Berzils** [comment]
Dear colleagues,

Wishing you a Merry Christmas, I would like to share two comments. Firstly, it is evident that the ESS managers consider the package a low priority. Despite promises of an API, more than six months have passed without any developments. I propose two solutions:

1. Retrieve all available data from the portal, including versions, corrections, etc., and place them in a repository accessible to us. This aligns with the solution suggested by @LukasWallrich and supported by @djhurio (Sveiki, Mārtiņ!), who stated, "(...) there should be a process to monitor if new data edition (release.version) for each round has been published by the ESS." I'm for this, but unfortunately I don't know how to do that. Otherwise, I would have done it already. 

2. Download the entire ESS to your computer. This is what I have done, and the file size is approximately 200 MB when zipped.

Best regards,

Jānis

**Comment by DaniMori** [comment]
Progress on the "data access" revamping issue has been dead for more than a year, for what I can see. Is that the case?

I assume you came to a dead end regarding this and no one had the time to put effort into it. I might be interested of the developments, and might even get some free time to contribute during the summer, but I'd like to know if it's still viable to get it to work.

Thanks a lot!

**Comment by briatte** [comment]
Hi everyone,

FWIW, I've [emailed Sikt](https://sikt.no/en/om-sikt/contact-us) to ask if what we're trying to do is even doable from their perspective. It's very possible that the platform has not been built in a way that would allow the queries to work in a systematic way.

**Comment by DaniMori** [comment]
Oh, seems like the only way then to interact now with the ESS is by manually downloading the datasets of interest.
That's too bad for the sake of reproducibility :(

**Comment by DaniMori** [comment]
Btw one of their claims in [the blog post where they announced the changes to data access](https://www.europeansocialsurvey.org/news/article/way-you-access-our-data-changing) says that

> This new cloud-based platform will also enable increased ability to integrate with relevant European/global services such as EduGAIN, DataCite and the European Open Science Cloud (EOSC).

So I would understand that they have plans to release a web service at some point.

**Comment by briatte** [comment]
Hey all,

Encouraging news below:

> Dear François,
>  
> Thank you for reaching out and sharing your efforts to access ESS data programmatically through GraphQL queries. We appreciate the ongoing interest from the academic community in utilizing ESS data efficiently.
>  
> We currently have a team working on developing an API that will allow users to retrieve metadata and data. We are keen to involve users like yourselves to ensure that the solution meets user needs and undergoes thorough testing.
> 
> We would be pleased to invite you for a meeting to discuss this further and explore collaboration opportunities. Please feel free to keep in touch with ---- (in cc.)  for more information and to schedule a meeting.
>  
> Thank you again for your initiative, and we look forward to working together.

I'll happily attend any online meeting. Feel free to email me if you want me to include you in my reply to the ESS team (`f.` + my GitHub username + `gmail.com`).

**Comment by patcon** [comment]
Any update? Thanks for all the energy everyone!

**Comment by DaniMori** [comment]
> Any update? Thanks for all the energy everyone!

Hi! @briatte CC'd me in the emails with the development team at ESS. What I can tell you is that they replied with instructions for querying the new API. After some initial tests, I can only say that progress in this respect is stale for now... 🙁

I'd really love to put some effort into it, but right now I have it on "low priority" myself.

**Comment by patcon** [comment]
Thanks 🙏🏻  I totally understand constraints of your attention.

In the meantime, would it be possible to copy the emailed instructions into a gist.github.com and link it here?

**Comment by eirik-sikt** [comment]
Hi all! 

The ESS API is now publicly available at https://api.ess.sikt.no/docs. It is very much still in beta, and any feedback can be sent to essdatasupport@sikt.no (or you can just share it here).

At the moment the API only supports data file downloads, but we are hoping to have some endpoints for metadata soon.

Best regards,
Eirik (working for the ESS data archive at Sikt)

**Comment by etiennebacher** [comment]
Hi all, I won't have the bandwidth to develop the new version of the package or to maintain it, but I have created a basic foundation using `httr2` if anyone wants to take it and start from there. Maybe I'll be able to contribute here and there but not more than that (also because I don't use this API in my daily work anymore):

https://github.com/etiennebacher/essurvey

Short example:

``` r
library(essurvey)

# I have already added an environment variable "ESS_API_KEY" with my key

get_ess_link(doi_prefix = "10.21338", doi_suffix = "ess6e02_6")
#> https://api.ess.sikt.no/v1/data/dataFile/10.21338/ess6e02_6/?userId=<user-id>&fileFormat=parquet

get_ess_data(doi_prefix = "10.21338", doi_suffix = "ess6e02_6") |> 
  tibble::as_tibble()
#> # A tibble: 54,673 × 625
#>    name   essround edition proddate  idno cntry dweight pspwght pweight anweight
#>    <chr>     <int> <chr>   <chr>    <int> <chr>   <dbl>   <dbl>   <dbl>    <dbl>
#>  1 ESS6e…        6 2.6     23.11.2…     1 AL     0.775   0.536    0.186   0.0997
#>  2 ESS6e…        6 2.6     23.11.2…     2 AL     1.35    0.660    0.186   0.123 
#>  3 ESS6e…        6 2.6     23.11.2…     3 AL     2.93    1.82     0.186   0.338 
#>  4 ESS6e…        6 2.6     23.11.2…     4 AL     0.969   0.534    0.186   0.0993
#>  5 ESS6e…        6 2.6     23.11.2…     5 AL     1.33    1.88     0.186   0.350 
#>  6 ESS6e…        6 2.6     23.11.2…     6 AL     0.0481  0.0683   0.186   0.0127
#>  7 ESS6e…        6 2.6     23.11.2…     7 AL     1.27    1.90     0.186   0.354 
#>  8 ESS6e…        6 2.6     23.11.2…     8 AL     1.33    1.95     0.186   0.362 
#>  9 ESS6e…        6 2.6     23.11.2…     9 AL     0.952   1.16     0.186   0.217 
#> 10 ESS6e…        6 2.6     23.11.2…    10 AL     0.772   0.454    0.186   0.0845
#> # ℹ 54,663 more rows
#> # ℹ 615 more variables: tvtot <int>, tvpol <int>, ppltrst <int>, pplfair <int>,
#> #   pplhlp <int>, polintr <int>, trstlgl <int>, trstplc <int>, trstplt <int>,
#> #   trstep <int>, trstun <int>, trstprt <int>, trstprl <int>, vote <int>,
#> #   prtvtal <int>, prtvtcbe <int>, prtvtcbg <int>, prtvtdch <int>,
#> #   prtvtacy <int>, prtvtccz <int>, prtvdde1 <int>, prtvdde2 <int>,
#> #   prtvtcdk <int>, prtvtdee <int>, prtvtces <int>, prtvtcfi <int>, …
```

**Comment by DaniMori** [comment]
> Hi all, I won't have the bandwidth to develop the new version of the package or to maintain it, but I have created a basic foundation using `httr2` if anyone wants to take it and start from there. Maybe I'll be able to contribute here and there but not more than that (also because I don't use this API in my daily work anymore):

@etiennebacher it would be advisable to avoid exposing your "user ID", IMHO. Anyone who wants to test the API can get theirs by logging in to the data portal and browsing to https://ess.sikt.no/en/api

Re. implementation, I asked the following questions in an email exchange with @briatte and @eirik-sikt, on 2025/05/08:

> @François Briatte, in terms of implementation, I guess the highest concern would be response times.
I'd suggest that maybe the query result were cached locally for the session. What do you guys think?
>
> I understand now, though (after going through the API docs a little), that there are several datasets, and it is possible to query all of them by using the DOI. The only "complication", I guess, is to translate a "plain text" description of a dataset to its DOI (depending on how many there are, but it could be just a lookup table?).

Do you have any idea of how these go, @etiennebacher ?

**Comment by etiennebacher** [comment]
Whoops, didn't notice it was in the output. Not a huge deal since the ESS API docs say:

>  In order to use the API to download data files, you must use your ESS user ID as an input parameter. Note that this parameter is not for authentication - it is used to track data usage statistics. 

But I'll edit this, thanks for noticing.

> > François Briatte, in terms of implementation, I guess the highest concern would be response times.
> > I'd suggest that maybe the query result were cached locally for the session. What do you guys think?
> >
> > I understand now, though (after going through the API docs a little), that there are several datasets, and it is possible to query all of them by using the DOI. The only "complication", I guess, is to translate a "plain text" description of a dataset to its DOI (depending on how many there are, but it could be just a lookup table?).
>
> Do you have any idea of how these go, @etiennebacher ?

`httr2` has [`req_cache()`](https://httr2.r-lib.org/reference/req_cache.html) that could probably be used for that. I never had to use this package much beyond the basics but it has quite a lot of built-in functionalities for all concerns regarding wrapping an API. There's also a [dedicated vignette](https://httr2.r-lib.org/articles/wrapping-apis.html) about wrapping APIs.

Re translating a plain text description to a DOI, I guess I would find it more useful to have a simple table on the ESS API website that lists the DOI and a description of their content. There could be a function to produce a data.frame out of this table, but I guess my own usecase would be to check manually on the website what DOIs I actually need and then read them with the package.

**Comment by DaniMori** [comment]
> Whoops, didn't notice it was in the output. Not a huge deal since the ESS API docs say:

Not a big deal in terms of authentication, you're right on that; it's just for the sake of not messing up with ESS's usage stats, but I'm sure they'll appreciate it 😄 

> `httr2` has [`req_cache()`](https://httr2.r-lib.org/reference/req_cache.html) that could probably be used for that. I never had to use this package much beyond the basics but it has quite a lot of built-in functionalities for all concerns regarding wrapping an API. There's also a [dedicated vignette](https://httr2.r-lib.org/articles/wrapping-apis.html) about wrapping APIs.

Great to know that! If I manage to get hands on this, I'll have a deeper look at it 😉 

> Re translating a plain text description to a DOI, I guess I would find it more useful to have a simple table on the ESS API website that lists the DOI and a description of their content. There could be a function to produce a data.frame out of this table, but I guess my own usecase would be to check manually on the website what DOIs I actually need and then read them with the package.

I was thiking of the current implementation of the package, where different datasets have different functions to help have a clear interface that allows easy understanding of what one is querying. I'm probably the farthest in expertise, but I was thiking of the `import_*()` family of functions. Does this make sense?

**Comment by DaniMori** [comment]
Btw, for the sake of order and "findability", I'd suggest closing this issue and continuing the discussion in #56.

**Comment by etiennebacher** [comment]
I think #56 should be closed instead (there's no information there that is also not present here) and this issue should be kept open until a maintainer is found.

**Comment by etiennebacher** [comment]
Since the API is in beta, we can list here what we would need to reproduce the functionalities of the defunct `essurvey`. 

I can't take a deep dive into it right now but I think the main issue for now is documentation. AFAICT, the only API pages are:

* https://ess.sikt.no/en/api, which just mentions the ESS User ID 
* https://api.ess.sikt.no/docs, which lists 5 params so far. I think we'd need to be able to pass the years and countries of interest in the query to reproduce the `import_*()` functions of the previous version.

Having a table about the list of available DOIs prefixes and suffixes is also essential.

cc @eirik-sikt

**Comment by DaniMori** [comment]
I created issues #59 and #60 to refer to (some of) the necessary updates. Maybe there should be a new issue (e.g. "Update access to ESS Data to new API interface"), with those (and more that come up along the way) as sub-issues?

**Comment by eirik-sikt** [comment]
> Having a table about the list of available DOIs prefixes and suffixes is also essential.

As mentioned, I am hoping to be able add some endpoints for fetching ESS metadata soon. Contained within that metadata would be an overview of datasets with their DOIs.

**Comment by patcon** [comment]
I think we can easily do this ourselves with a simple json or yaml file, and so you should not consider us blocked on your doing anything :) Should be easy enough to swap out whenever your solution is ready. thank you for the API so far!

**Comment by patcon** [comment]
In case it's helpful, I wrote a quick script for a best-effort scrape of all the datafile pages that archive.org is aware of:
https://gist.github.com/patcon/2129ffcff1d247af2a9cb645db66be1c

CSV found 115 files, but if my url pattern is wrong, someone should be able to run it again pretty easily and get a better list :) Took maybe an hour to write with chatgpt and 5 minutes to run.

EDIT: a quick check makes me think these are all working with the API. Someone can probably infer some missing ones that internet archive had never seen before, but usually the archive is good at combing scientific sites like this, since it's a big part of their mission

**Comment by maelle** [comment]
Does anyone need write access to this repository? 😸 

Very cool to see the productive discussion in any case. 👏

**Comment by maelle** [comment]
cc @ropensci/admin

**Comment by DaniMori** [comment]
I may be able to contribute from time to time, thanks @maelle!

Right now I could contribute by helping manage issues (add labels, assignees, etc.), so more than write access, maybe maintainer permissions?

**Comment by maelle** [comment]
@DaniMori I gave you Write access (you need to accept an invitation that you got by email or through GitHub notification) which includes Triage access. Thank you!!

**Comment by DaniMori** [comment]
Done! I guess that means I'm a maintainer now 🙂 

Are there guidelines for labelling and triage?

**Comment by maelle** [comment]
No active maintainer, no rules yet 😸 We do use "help wanted" over the whole rOpenSci organization though: https://ropensci.org/help-wanted/

**Comment by DaniMori** [comment]
I made a little progress, up to a point where I came to the conclusion that a DOI-dataset dictionary is needed.

Right now the project where I needed this package is on pause, so I've had to lower the priority I'm afraid 🙁 

I may be able to find some time to put into this from november on though 🤞🏽

**Comment by maelle** [comment]
Thanks a lot for the update, fingers crossed! 🙏
